modules/parent-join/src/main/java/org/elasticsearch/join/mapper/ParentJoinFieldMapper.java

```diff
@@ -0,0 +1,407 @@
+                conflicts.add("cannot remove parent [" + mapper.getParentName() + "] in join field [" + name() + "]");
+            }
+        }
+
+        final ArrayList<ParentIDFieldMapper> newParentIDFields = new ArrayList<>();
+        for (ParentIDFieldMapper mergeWithMapper : joinMergeWith.parentIDFields) {
+            ParentIDFieldMapper self = getParentIDFieldMapper(mergeWithMapper.getParentName(), true);
+            if (self == null) {
+                if (getParentIDFieldMapper(mergeWithMapper.getParentName(), false) != null) {
+                    // it is forbidden to add a parent to an existing child
+                    conflicts.add("cannot create parent [" + mergeWithMapper.getParentName()  + "] from an existing child");
+                }
+                for (String child : mergeWithMapper.getChildren()) {
+                    if (getParentIDFieldMapper(child, true) != null) {
+                        // it is forbidden to add a parent to an existing child
+                        conflicts.add("cannot create child [" + child  + "] from an existing parent");
+                    }
+                }
+                newParentIDFields.add(mergeWithMapper);
+            } else {
+                for (String child : self.getChildren()) {
+                    if (mergeWithMapper.getChildren().contains(child) == false) {
+                        conflicts.add("cannot remove child [" + child + "] in join field [" + name() + "]");
+                    }
+                }
+                ParentIDFieldMapper merged = (ParentIDFieldMapper) self.merge(mergeWithMapper, false);
+                newParentIDFields.add(merged);
+            }
+        }
+        if (conflicts.isEmpty() == false) {
+            throw new IllegalStateException("invalid update for join field [" + name() + "]:\n" + conflicts.toString());
+        }
+        this.parentIDFields = Collections.unmodifiableList(newParentIDFields);
+    }
+
+    @Override
+    public FieldMapper updateFieldType(Map<String, MappedFieldType> fullNameToFieldType) {
+        ParentJoinFieldMapper fieldMapper = (ParentJoinFieldMapper) super.updateFieldType(fullNameToFieldType);
+        final List<ParentIDFieldMapper> newMappers = new ArrayList<> ();
+        for (ParentIDFieldMapper mapper : fieldMapper.parentIDFields) {
+            newMappers.add((ParentIDFieldMapper) mapper.updateFieldType(fullNameToFieldType));
+        }
+        fieldMapper.parentIDFields = Collections.unmodifiableList(newMappers);
+        return fieldMapper;
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context, List<IndexableField> fields) throws IOException {
+        throw new UnsupportedOperationException("parsing is implemented in parse(), this method should NEVER be called");
+    }
+
+    @Override
+    public Mapper parse(ParseContext context) throws IOException {
+        // Only one join field per document
+        checkDuplicateJoinFields(context.doc());
```

Comment:
Maybe we should only allow one join field definition per index? (not sure what the best way to validate this)
Because all parent/child relations can be expressed in a single join field.

Also as is now one is free to add a join field mapper to an index, which I think is problematic when documents have already been indexed.